### DISCUSSION TITLE
Obtaining a reference to media uploaded via SpatieMediaLibraryFileUpload

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#1604] [Help]
Just looking for a bit of guidance on how to get/set some kind of reference to a entry in the media table from the state when uploading media via `SpatieMediaLibraryFileUpload`.

When used, the media is pushed directly into the media table and we can obtain it using `getFirstMediaUrl`, For example.

```
SpatieMediaLibraryFileUpload::make('myfiles')
    ->collection('myfiles')
    ->mediaName('the_name_of_my_file')

//we can now use:
$post->getFirstMediaUrl('myfiles', 'the_name_of_my_file')
```
This is fine if we know ahead of time what collection and name the media is going to be associated with but for my use case, I want to use the field in a Builder Block which results in an issue:

With the following:

```
Builder::make('content')
   ->blocks([
        Block::make('my_section')
            ->schema([
                SpatieMediaLibraryUpload::make('my_image')
                    ->collection('my_images')
                    ->mediaName('my_special_image')
            ])
    ])
```

If I create one block and save, when I create a new block, the second block will have the image specified in the first. This is because Filament is pulling media named `my_special_image` from collection `my_images`. Even when using closures for either `collection` or `mediaName`, I couldn't see a way obtaining some kind of unique reference to be used to ensure these are unique.

Ideally, I would be able to set the collection and name dynamically and when I iterate through my array of blocks I can construct dynamic arguments for `getFirstMediaUrl`.

Any advice?

As an aside, with FileUpload, this problem goes away since the name of the file is stored along with the state but we don't make use of the Media Library. Would an option to optionally store the UUID (or id) of the media entry into the state be a solution? Something like:

```
SpatieMediaLibraryUpload::make('my_image')
   ->linkUuidToState(); //stores the uuid of the media to field "my_image"?

// and/or

SpatieMediaLibraryUpload::make('my_image')
   ->linkIdToState(); //stores the id of the media to field "my_image"?
```

I'm just "spitballing" here! Any help/advice/"you're doing it completely wrong" appreciated!

**Comment by danharrin** [comment]
The media library component does store the UUID of the media while you are using it - but the difference is that the UUID is loaded automatically from the record when the form is first loaded, and since media library does not store which media belongs to which builder item, just that it belongs to the Eloquent record, Filament doesnt know what media should be loaded in which block.

So, each media item will need a unique identifier based on the builder item.

Maybe you could use the new `mediaName()` method to do something like `mediaName(fn ($component) => $component->getStatePath())`, or the same with `collection()` if you're dealing with multi uploads. Since the components state path will always be unique, this will probably solve your issue, I think?

**Comment by titantwentyone** [comment]
Thanks Dan. This works in that the `name` now stores the `statePath`. However in the Filament UI, if two or more blocks are created, the same image is displayed in multiple components. Also, when other images are added, other components are affected. For now, I'm going to just use FileUpload for now but I will come back to this and see what can be done. Just would be nice to leverage the MediaLibrary.

Just to add some context, I'm creating a very (read very!) basic "Page Builder" for a client (Normally I hate these things). They just add and remove `Block` components, change the order etc and long term, I want them to be able to manage all their media in one place hence wanting the integration with Spatie.

**Comment by kjlloydie** [comment]
I was also looking to use `SpatieMediaLibraryFileUpload` in Builder blocks (in a very similar use-case to @titantwentyone by the sound of it). I have the same issue with the same image being loaded for every builder block, even with setting the media name to the state path (presumably because media library doesn't store the media name/reference in the JSON for the block, just in the model as you said).

@danharrin so you have any plans to support using the plugin within Builder blocks, or should I just switch to using FileUpload too?

**Reply by danharrin** [reply]
I don't have plans to officially support it, since Spatie does not have plans to support attaching files to JSON array keys. If this changes, I am more than happy to reconsider, but I'd rather not have hacky workarounds within the Filament codebase.
